docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,421 @@
+
+The most favorable approach appears to be defining routing on the client side. Instead of creating separate cluster sets for each client on the Egress, we create an 'aggregate_cluster' for the client:
+
+Aggregate_cluster:
+
+- backend_in_zone
+- backend_zone1
+- backend_zone2
+
+Each entry represents the Envoy cluster name that the client is interested in communicating with, in order of priority. This arrangement enables each cluster to set a separate 'SNI' for each zone, facilitating matching and routing on the Egress side.
+
+For example:
+
+```yaml
+backend_zone1 = SNI backend{mesh=default,kuma.io/zone=zone1}
+backend_zone2 = SNI backend{mesh=default,kuma.io/zone=zone2}
+```
+
+In addition, it is important to configure the Egress correctly. For each `SNI`, we need to create a filter chain match that sets the required metadata for selecting endpoints:
+
+```yaml
+- filterChainMatch:
+    serverNames:
+    - backend{mesh=default,kuma.io/zone=zone1}
+    transportProtocol: tls
+  filters:
+  - name: envoy.filters.network.tcp_proxy
+    typedConfig:
+      '@type': type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
+      cluster: default:backend
+      metadataMatch:
+        filterMetadata:
+          envoy.lb:
+            kuma.io/zone: zone1
+      statPrefix: default:backend_zone1
+```
+Each dataplane has metadata, and it's important to ensure that `kuma.io/zone` is available. If it's not present, add it to the endpoints.
+
+On the ingress side, you only need to adjust the `SNI` to include its corresponding `kuma.io/zone`.
+
+
+##### Ingress
+We don't configure ingresses.
+
+#### ClusterLoadBalancing cache
+Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with localityAwareness load balancing because each dataplane might have a different subset of endpoints with different priorities.
+
+We are going to implement this without the cache and verify performance.
+
+Steps:
+* Run `mesh-perf` test to measure impact of not using CLA cache
+* If results are good we'll remove the CLA cache
+* Otherwise rewrite CLA cache to keep the different clusters depending on the dataplane tags
+
+##### Test result
+We have ran a test of 1000 services with 2 instances each in standalone mode. Each service had reachable services to 3 services defined.
+
+![With CLA Cache](assets/031/xds_generation_with_cache.png)
+
+![Without CLA Cache](assets/031/xds_generation_without_cache.png)
+
+We can observe that p90 and p99 are much higher without using the cache. We should keep records in the cache but we have to ensure that generated key is reusable so we don't create to many same CLA entries. Also, we could move CLA cache into MeshContext to keep related things together.
```

Comment:
We could add to the key string of priority to tag and value to make all services that would be in the same location and have the same policy to reuse CLA. Still, there is a chance that we might create much more in case there are many different specific client rules.

`0:k8s.io/node:value,1:k8s.io/az:value,value2,2:kuma.io/zone:zone1`

Reply:
By top targetRef you means 

```
targetRef:
  kind: Mesh
to:
  targetRef:
    kind: MeshService
    name: backend
```

or 
```
targetRef:
  kind: Mesh
```

Reply:
I don't think it's enough. Because we can have in `to` section `Mesh` and in another case `MeshService` which might provide a different configuration for the service.

Reply:
we will deal with CLA cache later on as a separate issue